extract_sets/training_sets.py

```diff
@@ -0,0 +1,398 @@
+                elem_id = self.element.attrib['id']
+                print("[ERROR  ] skip '{}': invalid coords!".format(
+                    elem_id), file=sys.stderr)
+                self.valid = False
+                return
+            texts.append(self.element)
+
+        sorted_els = sorted(
+            texts,
+            key=lambda w: int(PageLine._pick_top_left(w, self.namespace)))
+        unicodes = [
+            w.find(
+                f'.//{self.namespace}:Unicode',
+                XML_NS) for w in sorted_els]
+        self.text_words = [u.text.strip() for u in unicodes if u.text]
+
+        # elimiate read order mark
+        for i, strip in enumerate(self.text_words):
+            strip = self.text_words[i]
+            for mark in CLEAR_MARKS:
+                if mark in strip:
+                    self.text_words[i] = strip.replace(mark, '')
+
+    @staticmethod
+    def _pick_top_left(elem, namespace):
+        coords = elem.find(f'{namespace}:Coords', XML_NS)
+        points = coords.attrib['points'].split()
+        if len(points) > 0:
+            return points[0].split(',')[0]
+
+    def get_shape(self, element):
+        """
+        Coordinate data from current OCR-D-Workflows can contain
+        lots of points, therefore additional calculations are required
+        """
+
+        p_attr = element.find(
+            f'{self.namespace}:Coords',
+            XML_NS).attrib['points']
+        numbers = [int(n) for pair in p_attr.split() for n in pair.split(',')]
+
+        # group clustering idiom
+        points = list(zip(*[iter(numbers)] * 2))
+
+        return Polygon(points)
+
+
+def text_line_factory(xml_data, min_len, reorder):
+    """Create text_lines from given structured data"""
+
+    text_lines = []
+    ns_prefix = _determine_namespace(xml_data)
+    if 'alto' in ns_prefix:
+        all_lines = xml_data.findall(f'.//{ns_prefix}:TextLine', XML_NS)
+        all_lines_len = [l for l in all_lines if len(' '.join(
+            [s.attrib['CONTENT'] for s in l.findall(f'{ns_prefix}:String', XML_NS)])) >= min_len]
+        text_lines = [ALTOLine(line, ns_prefix) for line in all_lines_len]
+    elif ns_prefix in ('page2013', 'page2019'):
+        all_lines = xml_data.findall(f'.//{ns_prefix}:TextLine', XML_NS)
+        matchings = []
+        for textline in all_lines:
+            text_equiv = textline.find(
+                f'{ns_prefix}:TextEquiv/{ns_prefix}:Unicode', XML_NS)
+            if text_equiv.text:
+                stripped = text_equiv.text.strip()
+                if len(stripped) and len(stripped) >= min_len:
+                    matchings.append(textline)
+            else:
+                words = textline.findall(
+                    f'{ns_prefix}:Word/{ns_prefix}:TextEquiv/{ns_prefix}:Unicode', XML_NS)
+                if len(words):
+                    msg = f"[{xml_data.base}] no text but words for line '{textline.attrib['id']}'"
+                    raise RuntimeError(msg)
```

Comment:
Isn't this overly restrictive? You are preferring Word content in PageLine anyway, so why require TextEquiv on the TextLine level? You could just verify all the Word contents sum up to min_len here, like above.

Reply:
I've seen some samples of broken data, therefore the paranoia

Reply:
Broken in what way? PAGE-XML does permit a TextLine without TextEquiv having Words with TextEquiv. In the extraction, you even distrust the line-level text and concatenate from the word-level elements sorted by coordinates (which I agree is actually the right thing to do for OCR training, and much more robust than trusting `@readingDirection` and element ordering, especially with RTL and LTR/RTL mix). I guess it depends on the tool that produced the XML what you get.

Reply:
Yep, that's exactly the point. We are confronted (rather often) with so-called PAGE-XML from external sources produced by unknown Tools. It's rumored to be done with some Transkribus of yore, but it looks like a human fumbled straight into XML.

Reply:
But that's exactly why I am wondering here what makes you give up with an exception. Why write that robust Word/TextEquiv extractor if you are not using it when it's needed the most?

Reply:
The Problem is a *real* Problem just in rather exotic circumstances. We were confronted with Transkriptions in PAGE exported from Transkribus with only lines, no words, therefore missing proper word-coordinates for re-ordering.